### ISSUE TITLE
Security decorators (middleware) ?

### ISSUE THREAD

**Issue by iangregsondev** [smartive/giuseppe#41]
Hey, great library.

I was wondering if its possible to pass in custom middleware on a route or is it planned ?

or even better, protecting routes using security decorators ?

**Comment by buehler** [comment]
Hey @ianjohnspain 
great idea!

Some kind of a `@Security` or anything like that? Or what would you purpose? By now, we did not have any ideas about security filtering.

As of the actual version, you can use the returned router of the `registerControllers` function to add a middleware.

**Comment by iangregsondev** [comment]
Hey @buehler, 

Yes I was thinking something along those lines. Making taking ideas from other implementation i.e. Spring Security or similar.

There is so many types of security implementation, I think OAUTH would  be good.

I didn't realise you could add middleware on the registerControllers, its there an example somewhere ?

**Comment by buehler** [comment]
Hey @ianjohnspain,
Alright, I'll look into those frameworks to see what they do. Well for the general authentication process, I'd use [passport](http://passportjs.org/). My general idea of those security decorators would be some kind of a auth-predicate like:

``` typescript
@Security(req => req.user.role === 'ADMIN')
public foobar(): void{}
```

and to infuse the `user`, I'd use passport. Passport has great auth mechanisms for nearly every auth possible.

Actually there is no example for registering middlewares. But good point :wink:

The `registerControllers` function returns a expressJS router. So you can just use this router to add your middlewares. Or create the router firsthand and pass it to `registerControllers`.

e.g.

``` typescript
let router = express.Router();

router.use(passport);
router.use(/* whatever middleware */);

registerControllers('baseUrl', router);

app.use(router);
```

Cheers

**Comment by iangregsondev** [comment]
@buehler, I like the idea of passport!

Thanks for the example of register controllers.

**Comment by buehler** [comment]
hey @ianjohnspain 
A PR is open with the possibility to add middleware directly to the whole controller or to a route :)
it'll be in the next release of giuseppe!

Cheers

**Comment by iangregsondev** [comment]
Hi @buehler ,

Yep I just saw that, thats great news! Thanks for keeping me informed.

**Comment by buehler** [comment]
@ianjohnspain we released v0.6 today :) check it out: https://github.com/smartive/giuseppe/releases/tag/v0.6.0
